lib/pbench/server/cache_manager.py

```diff
@@ -430,29 +429,61 @@ def get_info(self, path: Path) -> JSONOBJECT:
         return fd_info
 
     @staticmethod
-    def extract(tarball_path: Path, path: str) -> Optional[str]:
-        """Extract a file from the tarball and return it as a string
+    def filestream(tarball_path: Path, path: Path) -> Optional[IO[bytes]]:
+        """Extracts filestream from the tarball
+
+        Reports failure by raising Exceptions
+
+        Args:
+            path: relative path within the tarball
+
+        Raise:
+            BadDirpath on failure extracting the file from tarball
+        """
+        try:
+            return tarfile.open(tarball_path, "r:*").extractfile(path)
+        except Exception as exc:
+            raise BadDirpath(
+                f"A problem occurred processing {path!r} from {tarball_path!s}: {exc}"
+            ) from exc
+
+    @staticmethod
+    def filecontents(tarball_path: Path, path: Path) -> Optional[str]:
+        """Returns filestream as a string"""
+        return Tarball.filestream(tarball_path, path).read().decode()
 
-        Report failures by raising exceptions.
+    def extract(self, path: str) -> Optional[tuple[dict, IO[bytes]]]:
+        """Extracts the contents of a file from the Tarball
 
         Args:
             path: relative path within the tarball of a file
 
         Raises:
-            MetadataError on failure opening the tarball
             TarballUnpackError on failure to extract the named path
 
         Returns:
-            The named file as a string
+            Tuple with file information and contents/stream
         """
+        name = Path(self.name)
+        path = name / path
+        info = self.get_info(path)
+
+        # Fix: Currently if target is not provided we are returning tarball.
+        # Do we want to do the same or there is an alternative approach that
+        # we can just return the info
```

Comment:
The cache manager owns and encapsulates access to the tarball, because only the cache manager knows where it is and how to access it. Whether we manage the special case here (return the tarball stream vs opening the tarball and returning the stream of a path within it) or require the caller to know and make a different call is the important question.

There's maybe some benefit to hiding the difference from `inventory` and other callers; but that doesn't mean it needs to be here, and "extract" is arguably an "inexact" name for a method that returns the tarball file stream. (We're not "extracting" anything.)